nautobot/extras/utils.py

```diff
@@ -207,6 +219,78 @@ def wrapper(model_class):
     return wrapper
 
 
+def populate_model_features_registry():
+    """
+    Populate the registry model features with new apps.
+
+    This function updates the registry model features.
+
+    Behavior:
+    - Defines a list of dictionaries called lookup_confs. Each dictionary contains:
+        - 'feature_name': The name of the feature to be updated in the registry.
+        - 'field_names': A list of names of fields that must be present in order for the model to be considered
+                        a valid model_feature.
+        - 'field_attributes': Optional dictionary of attributes to filter the fields by. Only model which fields match
+                            all the attributes specified in the dictionary will be considered. This parameter can be
+                            useful to narrow down the search for fields that match certain criteria. For example, if
+                            `field_attributes` is set to {"related_model": RelationshipAssociation}, only fields with
+                            a related model of RelationshipAssociation will be considered.
+    - Looks up all the models in the installed apps.
+    - For each dictionary in lookup_confs, calls lookup_by_field() function to look for all models that have fields with the names given in the dictionary.
+    - Groups the results by app and updates the registry model features for each app.
+    """
+    RelationshipAssociation = apps.get_model(app_label="extras", model_name="relationshipassociation")
+
+    lookup_confs = [
+        {
+            "feature_name": "custom_fields",
+            "field_names": ["_custom_field_data"],
+        },
+        {
+            "feature_name": "relationships",
+            "field_names": ["source_for_associations", "destination_for_associations"],
+            "field_attributes": {"related_model": RelationshipAssociation},
+        },
+    ]
+
+    app_models = apps.get_models()
+    for lookup_conf in lookup_confs:
+        registry_items = find_models_with_matching_fields(
+            app_models=app_models,
+            field_names=lookup_conf["field_names"],
+            field_attributes=lookup_conf.get("field_attributes"),
+        )
+        feature_name = lookup_conf["feature_name"]
+        registry["model_features"][feature_name] = registry_items
+
+
+def find_models_with_matching_fields(app_models, field_names, field_attributes=None):
```

Comment:
This probably belongs in `nautobot/core/models/utils.py` alongside the `get_all_concrete_models` function being added by #3368.

More generally, I wonder whether we need to align the approach between these two PRs. This PR is using duck-typing to identify relevant models, while #3368 is mostly using class inheritance to identify its relevant models. Would it make sense to pick one or the other and use it the same for both features?

Reply:
I would support this approach because it takes into consideration models that do not derive from a `ModelMixin`. As far as I know, not all plugin developers would add `status`, `role`, and other features to their models using inheritance, unless we plan to enforce inheritance.